web/api: add search API endpoint (#18573)
prometheus/prometheus: docs/querying/api.md

  if iteration failed mid-stream after the first batch was sent.

```json
{"results":[{"name":"http_requests_total","type":"counter","help":"Total HTTP requests."}]}
{"status":"success","has_more":false}
```

If an error occurs **before** streaming starts, the API returns the usual
Prometheus JSON error object with a 4xx/5xx status code. If an error occurs
**after** streaming starts, the stream ends with an NDJSON error line in place
of the trailer.

Clients must tolerate an abrupt EOF without a trailer (for example, on
transport failure or server shutdown) and must ignore unknown fields in the
trailer for forward compatibility.

The `has_more` field in the trailer is informational only: this version of
the API does not provide a pagination cursor. To retrieve more results, raise
`limit` (subject to the operator-configured `--web.search.max-limit`) or narrow
the request via `match[]`. A future version of the API may add a cursor.

Common URL query parameters:

- `match[]=<series_selector>`: Repeated series selector used to scope the
  search. Optional.
- `search[]=<string>`: Repeated search string matched against names or values.
  Multiple values use OR semantics. Optional.

> - `fuzz_threshold=<number>`: Fuzzy threshold from 0 to 100. Optional. A value

Reviewer: I suggest documenting the default:
```suggestion
- `fuzz_threshold=<number>`: Fuzzy threshold from 0 to 100. Optional. Default
  is 0, the lowest fuzzy threshold.
```

  of 0 is the lowest fuzzy threshold.
- `fuzz_alg=<subsequence | jarowinkler>`: Matching algorithm. Optional. Default
  is `subsequence`.

> - `case_sensitive=<bool>`: Toggle case-sensitive matching. Optional.

Reviewer: ```suggestion
- `case_sensitive=<bool>`: Toggle case-sensitive matching. Optional. Default
  is `true`.
```

> - `sort_by=<string>`: Sort mode. Supported values depend on the endpoint.

Reviewer: ```suggestion
- `sort_by=<string>`: Sort mode. Supported values depend on the endpoint.
  Optional; if unset, results are returned in natural order.
```

- `sort_dir=<asc | dsc>`: Sort direction. Optional. Only valid with
  `sort_by=alpha`.

> - `include_score=<bool>`: Include the relevance score in each result. Optional.

Reviewer: ```suggestion
- `include_score=<bool>`: Include the relevance score in each result. Optional.
  Default is `false`.
```

- `start=<rfc3339 | unix_timestamp>`: Start timestamp. Optional.
- `end=<rfc3339 | unix_timestamp>`: End timestamp. Optional.
- `limit=<number>`: Maximum number of returned results. Optional. Default is
  100.
- `batch_size=<number>`: Preferred number of results per NDJSON batch.
  Optional. Default is 100.

The `start` and `end` parameters narrow results to the selected time window.
Results may include values from series active slightly outside that window,
because Prometheus stores data in fixed-size blocks (typically 2 hours each).

Additional parameters for `/api/v1/search/metric_names`:

> - `include_metadata=<bool>`: Include metric metadata in each result.

Reviewer: ```suggestion
- `include_metadata=<bool>`: Include metric metadata in each result. Optional.
  Default is `false`.
```

- `sort_by=<alpha | score>`

Additional parameters for `/api/v1/search/label_names`:

- `sort_by=<alpha | score>`

Additional parameters for `/api/v1/search/label_values`:

- `label=<label_name>`: Label name whose values should be searched. Required.
- `sort_by=<alpha | score>`

This example searches metric names for autocomplete:

```bash
curl -g 'http://localhost:9090/api/v1/search/metric_names?search[]=http_req&sort_by=score&include_metadata=true&limit=5'
```

```json
{"results":[{"name":"http_requests_total","type":"counter","help":"Total HTTP requests."}]}
{"status":"success","has_more":false}
```

This example searches label values for the `instance` label within the `up`
metric:

```bash
curl -g 'http://localhost:9090/api/v1/search/label_values?label=instance&match[]=up&search[]=909&sort_by=score'
```

```json

> {"results":[{"name":"localhost:9090"},{"name":"localhost:9091"}]}

Reviewer: See comment on the `searchLabelValueResult` type:
```suggestion
{"results":[{"value":"localhost:9090"},{"value":"localhost:9091"}]}
```

- `sort_dir=<asc | dsc>`: Sort direction. Optional. Only valid with
  `sort_by=alpha`.
- `include_score=<bool>`: Include the relevance score in each result. Optional.
- `start=<rfc3339 | unix_timestamp>`: Start timestamp. Optional.
- `end=<rfc3339 | unix_timestamp>`: End timestamp. Optional.
- `limit=<number>`: Maximum number of returned results. Optional. Default is
  100.
- `batch_size=<number>`: Preferred number of results per NDJSON batch.
  Optional. Default is 100.

The `start` and `end` parameters narrow results to the selected time window.
Results may include values from series active slightly outside that window,
because Prometheus stores data in fixed-size blocks (typically 2 hours each).

Additional parameters for `/api/v1/search/metric_names`:

- `include_metadata=<bool>`: Include metric metadata in each result.
- `sort_by=<alpha | score>`

Additional parameters for `/api/v1/search/label_names`:

- `sort_by=<alpha | score>`

Additional parameters for `/api/v1/search/label_values`:

- `label=<label_name>`: Label name whose values should be searched. Required.
- `sort_by=<alpha | score>`

This example searches metric names for autocomplete:

```bash

> curl -g 'http://localhost:9090/api/v1/search/metric_names?search[]=http_req&sort_by=score&include_metadata=true&limit=5'

Reviewer 2: Is it worth showing an example with annotations/warnings in the result?

Author: I think this is too niche for now, I would skip this.
